crates/subspace-archiving/src/reconstructor.rs

```diff
@@ -0,0 +1,246 @@
+use crate::archiver::{Segment, SegmentItem};
+use crate::utils;
+use parity_scale_codec::Decode;
+use reed_solomon_erasure::galois_16::ReedSolomon;
+use std::mem;
+use subspace_core_primitives::{
+    ArchivedBlockProgress, LastArchivedBlock, Piece, RootBlock, PIECE_SIZE, SHA256_HASH_SIZE,
+};
+use thiserror::Error;
+
+/// Reconstructor-related instantiation error.
+#[derive(Debug, Error, Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash)]
+pub enum ReconstructorInstantiationError {
+    /// Segment size is not bigger than record size
+    #[error("Segment size is not bigger than record size")]
+    SegmentSizeTooSmall,
+    /// Segment size is not a multiple of record size
+    #[error("Segment size is not a multiple of record size")]
+    SegmentSizesNotMultipleOfRecordSize,
+    /// Wrong record and segment size, it will not be possible to produce pieces
+    #[error("Wrong record and segment size, it will not be possible to produce pieces")]
+    WrongRecordAndSegmentCombination,
+}
+
+/// Reconstructor-related instantiation error
+#[derive(Debug, Error, Clone, PartialEq)]
+pub enum ReconstructorError {
```

Comment:
Why `InstantiationError` is not part of `Error`? In this way, we have to deal with two error types from the reconstructor.